### PR TITLE
Fix for mean functions

### PR DESCRIPTION
**Pull request by bwengals**
This PR addresses #2078 by fixing the shape of mean function outputs.  It also contains tests for the bug in #2078.  I had tried including some refactoring but removed this.  Now this is just a fix for the mean functions + tests.

### CODE REVIEW

File: pymc3/gp/cov.py [hunk 1]
```diff
@@ -342,7 +337,7 @@ def square_dist(self, X, Z):
             sqd = -2.0 * tt.dot(X, tt.transpose(X)) +\
                   (tt.reshape(Xs, (-1, 1)) + tt.reshape(Xs, (1, -1)))
         else:
-            Z = tt.as_tensor_variable(Z)
+            Z = tt.as_tensor_variable(X)
```
Comment (twiecki) [inline pymc3/gp/cov.py]: Is this intentional? If so, might need a comment.

File: pymc3/gp/mean.py [hunk 2]
```diff
@@ -22,49 +71,51 @@ def __add__(self, other):
...
+
         Parameters
         ----------
         coeffs : variables
             Linear coefficients
         intercept : variable, array or integer
             Intercept for linear function (Defaults to zero)
         """
-        Mean.__init__(self)
+        Mean.__init__(self, input_dim, active_dims)
```
Comment (twiecki) [inline pymc3/gp/mean.py]: I think we should use `super` here.

File: pymc3/gp/cov.py (hunk 1 above)
Comment (bwengals) [reply pymc3/gp/cov.py]: no not intentional! good catch...

File: pymc3/gp/mean.py (hunk 2 above)
Comment (bwengals) [reply pymc3/gp/mean.py]: sounds good, fixed

**Comment by bwengals** [comment]
While addressing your corrections I found that `Add` and `Prod` for mean functions don't work correctly.  Will address in separate PR.  

**Edit:** I think this is due to my changes --- subclassing `Mean` on `ParameterizedFunction` which requires the `input_dim` and `active_dim` args.  Maybe this PR is the place for these fixes then.

File: pymc3/gp/gp.py [hunk 3]
```diff
@@ -13,49 +13,51 @@
...
-        
+
+        if mean_func is None:
+            self.M = Zero(input_dim=cov_func.active_dims)
+        else:
+            if not isinstance(mean_func, Mean):
+                raise ValueError('mean_func must be a subclass of Mean')
+            self.M = mean_func
+
+
```
Comment (fonnesbeck) [inline pymc3/gp/gp.py:47]: Was moving this an aesthetic change, or is there a reason for checking the covariance function first?

File: pymc3/gp/mean.py [hunk 4]
```diff
@@ -1,12 +1,61 @@
...
+    Parameters
+    ----------
+    input_dim : integer
+        The number of input dimensions, or columns of X (or Z) the kernel will operate on.
+    active_dims : A list of booleans whose length equals input_dim.
+	The booleans indicate whether or not the covariance function operates
+        over that dimension or column of X.
+    """
+
+    def __init__(self, input_dim, active_dims=None):
```
Comment (fonnesbeck) [inline pymc3/gp/mean.py]: Is there a downside to having a default of `input_dim=1`?

File: pymc3/gp/gp.py (hunk 3 above)
Comment (bwengals) [reply pymc3/gp/gp.py:47]: It was so the default mean function could take parameters that the user would have set when they instantiated the covariance function

File: pymc3/gp/mean.py (hunk 4 above)
Comment (bwengals) [reply pymc3/gp/mean.py]: hmm I'm not sure.  It's certainly the most likely thing a user would do.  When these functions are instantiated they don't know about `X` yet, so if a user forgets to set it explicitly they'll get an error at runtime.

**Comment by bwengals** [comment]
Also, are there other mean functions that might be of interest @fonnesbeck ?

**Comment by fonnesbeck** [comment]
I'm sure there are fancy things you can do with mean functions, but you don't really see anyone doing it. As long as we can apply a little prior information to the mean (which we can with a linear mean) then we are probably set for the time being.

**Comment by bwengals** [comment]
I decided to just address #2078, which was just a simple fix for outputs with the wrong shape.  I removed the scope creep...  Nother day.  I think this should be ready for review then @fonnesbeck @twiecki

**Comment by fonnesbeck** [comment]
These look good. Is this still a WIP, or are you ready to merge?

**Comment by bwengals** [comment]
Yes I think so @fonnesbeck.  Those 3 recent commits change the output shape for the mean functions from (n, 1) to (n, ), as per #2262 .

**Comment by aseyboldt** [comment]
@bwengals Needs a rebase.

**Comment by bwengals** [comment]
@aseyboldt resolved it

**Comment by aseyboldt** [comment]
The failing tests seem to be due to the numpy update to 1.13
We should probably fix that in a different PR, merge that and then rerun the tests.

**Comment by twiecki** [comment]
@fonnesbeck Should we merge this before 3.1?

**Comment by bwengals** [comment]
see #2337

**Comment by junpenglao** [comment]
close in favor of #2337

**Pull request closed without merging**